srml/contract/src/wasm/mod.rs

```diff
@@ -17,154 +17,166 @@
-	InternalMemoryDeclared,
+pub struct WasmExecutable {
+	entrypoint_name: &'static [u8],
+	prefab_module: PrefabWasmModule,
+}
 
-	/// Gas instrumentation failed.
-	///
-	/// This most likely indicates the module isn't valid.
-	GasInstrumentation,
+pub struct WasmLoader<'a, T: Trait> {
+	schedule: &'a Schedule<T::Gas>,
+}
 
-	/// Stack instrumentation failed.
-	///
-	/// This  most likely indicates the module isn't valid.
-	StackHeightInstrumentation,
+impl<'a, T: Trait> WasmLoader<'a, T> {
+	pub fn new(schedule: &'a Schedule<T::Gas>) -> Self {
+		WasmLoader { schedule }
+	}
+}
 
-	/// Error happened during invocation of the contract's entrypoint.
-	///
-	/// Most likely because of trap.
-	Invoke,
+impl<'a, T: Trait> ::exec::Loader<T> for WasmLoader<'a, T> {
+	type Executable = WasmExecutable;
 
-	/// Error happened during instantiation.
-	///
-	/// This might indicate that `start` function trapped, or module isn't
-	/// instantiable and/or unlinkable.
-	Instantiate,
+	fn load_init(&self, code_hash: &CodeHash<T>) -> Result<WasmExecutable, &'static str> {
+		let prefab_module = code::load::<T>(code_hash, self.schedule)?;
+		Ok(WasmExecutable {
+			entrypoint_name: b"deploy",
+			prefab_module,
+		})
+	}
+	fn load_main(&self, code_hash: &CodeHash<T>) -> Result<WasmExecutable, &'static str> {
+		let prefab_module = code::load::<T>(code_hash, self.schedule)?;
+		Ok(WasmExecutable {
+			entrypoint_name: b"call",
+			prefab_module,
+		})
+	}
+}
 
-	/// Memory creation error.
-	///
-	/// This might happen when the memory import has invalid descriptor or
-	/// requested too much resources.
-	Memory,
+pub struct WasmVm<'a, T: Trait> {
+	schedule: &'a Schedule<T::Gas>,
 }
 
-/// Execute the given code as a contract.
-pub fn execute<'a, E: Ext>(
-	code: &[u8],
-	input_data: &[u8],
-	output_data: &mut Vec<u8>,
-	ext: &'a mut E,
-	schedule: &Schedule<<E::T as Trait>::Gas>,
-	gas_meter: &mut GasMeter<E::T>,
-) -> Result<(), Error> {
-	let env = runtime::init_env();
-
-	let PreparedContract {
-		instrumented_code,
-		memory,
-	} = prepare_contract(code, &schedule, &env)?;
-
-	let mut imports = sandbox::EnvironmentDefinitionBuilder::new();
-	for (func_name, ext_func) in &env.funcs {
-		imports.add_host_func("env", &func_name[..], ext_func.raw_fn_ptr());
+impl<'a, T: Trait> WasmVm<'a, T> {
+	pub fn new(schedule: &'a Schedule<T::Gas>) -> Self {
+		WasmVm { schedule }
 	}
-	imports.add_memory("env", "memory", memory.clone());
+}
 
-	let mut runtime = Runtime::new(ext, input_data, output_data, &schedule, memory, gas_meter);
+impl<'a, T: Trait> ::exec::Vm<T> for WasmVm<'a, T> {
+	type Executable = WasmExecutable;
+
+	fn execute<E: Ext<T = T>>(
+		&self,
+		exec: &WasmExecutable,
+		ext: &mut E,
+		input_data: &[u8],
+		empty_output_buf: EmptyOutputBuf,
+		gas_meter: &mut GasMeter<E::T>,
+	) -> VmExecResult {
+		let memory =
+			sandbox::Memory::new(exec.prefab_module.initial, Some(exec.prefab_module.maximum))
+				.unwrap_or_else(|_| {
```

Comment:
why `unwrap_or_else(||{panic(...` rather than `.expect("...`?

Reply:
This is a trick which equivalent to `.expect` but preserves the source location! Super useful in contexts where you can't use `RUST_BACKTRACE=1`.
I also was reluctant to leave it in the PR because I was afraid that it can raise questions... (

Reply:
Good reasoning. How about adding a comment to prevent this from accidentally replaced then?
```suggestion
				.unwrap_or_else(|_| { // unlike `.expect`, explicit panic preserves the source location. Needed as we can't use `RUST_BACKTRACE` in here.
```

Reply:
Yeah, that's a good idea!

I also thought that maybe it is worth introducing a special macro, bike-sheddingly called, `expect_done_right!()` which just wraps panic and has a comment why we did that!